experimental/Schemes/WeilDivisor.jl

```diff
@@ -210,34 +210,49 @@ function intersect(D::WeilDivisor, E::WeilDivisor)
   return result
 end
 
-function colength(I::IdealSheaf)
+function colength(I::IdealSheaf; covering::Covering=default_covering(scheme(I)))
   X = scheme(I)
-  C = default_covering(X)
   patches_todo = copy(affine_charts(X))
   patches_done = AbsSpec[]
   result = 0
   while length(patches_todo) != 0
     U = pop!(patches_todo)
     J = I(U)
-    # To avoid overcounting, throw away all components that 
-    # were already visible in other charts.
-    for V in patches_done
-      if !haskey(glueings(C), (U, V))
-        continue
+    if has_decomposition_info(covering)
+      h = decomposition_info(covering)[U]
+      # The elements in h indicate where components must 
+      # be located so that they can not be spotted in other charts.
+      # We iteratively single out these components by adding a sufficiently high 
+      # power of the equation to the ideal.
+      for f in h
+        k = 0
+        while !(f^(k) in ideal(OO(U), f^(k+1)) + J)
+          k = k + 1
+        end
+        J = J + ideal(OO(U), f^k)
```

Comment:
Do you need the exakt value of k? I do not think so. Hence you might want to keep squaring the polynomial instead of multiplying f to it in each pass of the loop. This avoids a lot of standard basis computations hidden in the ideal membership test.